### ISSUE TITLE
auth-scheme missing from Authorization header when talking to registry

### ISSUE THREAD

**Issue by sfackler** [rust-lang/cargo#4989]
We currently generate a header that looks like `Authorization: asdfasdfasdfasdf`: https://github.com/rust-lang/cargo/blob/868d37339117213fdef4ead8daf4a9557cc4ec89/src/crates-io/lib.rs#L172

This does not match up with the grammar specified in [RFC 7235](https://tools.ietf.org/html/rfc7235#section-2.1), however: 

```
credentials = auth-scheme [ 1*SP ( token68 / #auth-param ) ]
Authorization = credentials
```

**Comment by sfackler** [comment]
If we change this for crates.io, we should make sure to allow custom registry tokens to be formatted however the registry requires - `Basic asdfasdfasdfsf` or `Bearer asdfasdfasdf` for example.

**Comment by TimDiekmann** [comment]
Wouldn't this be a breaking change as custom registries already landed?

**Comment by dtretyakov** [comment]
@ehuss  any updates on it? It seems that `crates` should follow standards.

What is your recommendation for custom registry vendors on that regard?

**Comment by dtretyakov** [comment]
Currently it's in https://github.com/rust-lang/cargo/blob/f1dadb6309e7d556f46ee272d63f522bbcf029bf/src/cargo/sources/registry/http_remote.rs#L591

**Comment by Eh2406** [comment]
There is basically no update.
The adherence to standards about the authorization header are lax in the real world. There are registries that require nonstandard authorization formats. The fact that cargo can send nonstandard authorization headers is a feature removing it would be a breaking change. It is also widely enough used that that breaking change is probably not going to happen. The confusing naming where the configuration `token` is in fact used as `authorization header` verbatim is unfortunate, but not easily changed.

Unfortunately the sparse protocol does not provide an opportunity for breaking changes, because the sparse protocol describes how to interact with the index and the authorization headers are currently sent to API endpoints. It is possible to have both indexes using GIT and sparse protocols that point to the same API endpoints.

There is also work about adding `auth-required`, which would allow sending the authorization headers to the index requests when using the sparse registry. So far that work is not included requiring compliance with the standards. Of course such a system would need some way to specify that you want to send a noncompliant header in case the registry needed that.

(P.S. you can edit your comments if you have more to add on the subject.)

**Comment by epage** [comment]
@arlosi wanted to check if any of the auth work affected this.

**Comment by arlosi** [comment]
> @arlosi wanted to check if any of the auth work affected this.

No. This is basically a crates.io issue. The "token" is an Authorization header, so in order to be compliant, it should start with `Bearer `, `Basic ` or similar. The naming "token" is a bit unfortunate and confusing. 

3rd party registries usually document how to configure the token appropriately. Everything eh2406 said above is still current.

**Comment by Turbo87** [comment]
FWIW with https://github.com/rust-lang/crates.io/pull/11350 crates.io now supports both empty and Bearer auth schemes (not deployed yet though)

**Comment by ehuss** [comment]
The Cargo team discussed this in https://github.com/rust-lang/cargo-team/blob/main/meetings/sync-meeting/2025-12-16.md#backlog-auth-scheme-missing-from-authorization-header-when-talking-to-registry.

We didn't really have an answer on what to do here, since we couldn't really think of anything that wouldn't be a breaking change. We discussed some options:

- The config file could say what format the registry wants, but that felt very unsatisfying.
- The credential API could be extended to support multiple auth schemes, but that doesn't really help with the common cases unless the auth helper made the user choose which scheme, which is also not satisfying.
- We could special-case crates.io as a short-term solution. We weren't sure what kinds of consequences that would have, and whether that would be worth adding as a special case.

I don't know if there could be other potential solutions. Could cargo maybe try both formats if the first one fails? Are there other ideas for how we could make this transition?

**Comment by Turbo87** [comment]
> The config file could say what format the registry wants, but that felt very unsatisfying.

```toml
[registries.foo]
index = "https://github.com/foo/foo-index"
token-scheme = "Bearer" # or `false`
```

I assume that means something like this? Could you clarify what feels "unsatisfying" about it?
